Data Loading and Preprocessing

In [1]:
import tensorflow as tf
from tensorflow.keras.datasets import cifar10
from tensorflow.keras.utils import to_categorical

# Load CIFAR-10 dataset
(x_train, y_train), (x_test, y_test) = cifar10.load_data()

# Normalize the data
x_train = x_train.astype('float32') / 255.0
x_test = x_test.astype('float32') / 255.0

# Convert labels to one-hot encoding
y_train = to_categorical(y_train, 10)
y_test = to_categorical(y_test, 10)


170498071/170498071 [==============================] - 4s 0us/step


Data Augmentation

In [2]:
from tensorflow.keras.preprocessing.image import ImageDataGenerator

datagen = ImageDataGenerator(
    rotation_range=15,
    width_shift_range=0.1,
    height_shift_range=0.1,
    horizontal_flip=True
)

datagen.fit(x_train)


Model Architecture

In [4]:
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Conv2D, MaxPooling2D, Flatten, Dense, Dropout, BatchNormalization

model = Sequential([
    Conv2D(32, (3, 3), activation='relu', padding='same', input_shape=(32, 32, 3)),
    BatchNormalization(),
    Conv2D(32, (3, 3), activation='relu', padding='same'),
    BatchNormalization(),
    MaxPooling2D(pool_size=(2, 2)),
    Dropout(0.25),

    Conv2D(64, (3, 3), activation='relu', padding='same'),
    BatchNormalization(),
    Conv2D(64, (3, 3), activation='relu', padding='same'),
    BatchNormalization(),
    MaxPooling2D(pool_size=(2, 2)),
    Dropout(0.25),

    Conv2D(128, (3, 3), activation='relu', padding='same'),
    BatchNormalization(),
    Conv2D(128, (3, 3), activation='relu', padding='same'),
    BatchNormalization(),
    MaxPooling2D(pool_size=(2, 2)),
    Dropout(0.25),

    Flatten(),
    Dense(512, activation='relu'),
    BatchNormalization(),
    Dropout(0.5),
    Dense(10, activation='softmax')
])

model.compile(optimizer='adam', loss='categorical_crossentropy', metrics=['accuracy'])



Training the Model

In [5]:
batch_size = 64
epochs = 50

history = model.fit(datagen.flow(x_train, y_train, batch_size=batch_size),
                    steps_per_epoch=x_train.shape[0] // batch_size,
                    epochs=epochs,
                    validation_data=(x_test, y_test))


Epoch 1/50
781/781 [==============================] - 47s 46ms/step - loss: 1.7156 - accuracy: 0.4203 - val_loss: 1.4021 - val_accuracy: 0.5238
Epoch 2/50
781/781 [==============================] - 32s 41ms/step - loss: 1.1743 - accuracy: 0.5815 - val_loss: 1.4795 - val_accuracy: 0.5626
Epoch 3/50
781/781 [==============================] - 34s 43ms/step - loss: 0.9974 - accuracy: 0.6471 - val_loss: 1.0007 - val_accuracy: 0.6606
Epoch 4/50
781/781 [==============================] - 33s 43ms/step - loss: 0.8986 - accuracy: 0.6840 - val_loss: 0.8323 - val_accuracy: 0.7196
Epoch 5/50
781/781 [==============================] - 33s 42ms/step - loss: 0.8419 - accuracy: 0.7052 - val_loss: 1.0191 - val_accuracy: 0.6624
Epoch 6/50
781/781 [==============================] - 34s 43ms/step - loss: 0.7833 - accuracy: 0.7277 - val_loss: 0.7349 - val_accuracy: 0.7459
Epoch 7/50
781/781 [==============================] - 33s 42ms/step - loss: 0.7361 - accuracy: 0.7437 - val_loss: 0.7141 - val_accuracy:

Model Evaluation

In [6]:
# Evaluate the model on the test set
test_loss, test_accuracy = model.evaluate(x_test, y_test)
print(f"Test Loss: {test_loss}")
print(f"Test Accuracy: {test_accuracy}")

313/313 [==============================] - 1s 4ms/step - loss: 0.3928 - accuracy: 0.8700
Test Loss: 0.39283937215805054
Test Accuracy: 0.8700000047683716


Test Loss: A test loss of 0.3928 indicates that the model is reasonably well-fitted to the test data. Loss values closer to zero are better, but this value suggests that the model is performing well, especially considering the complexity of the CIFAR-10 dataset.

Test Accuracy: An accuracy of 87.00% is a significant improvement. It indicates that the model correctly classifies 87 out of 100 images. This is a good result for CIFAR-10, where achieving very high accuracy is challenging due to the dataset's diversity and complexity.

Data Augmentation: Using data augmentation has likely contributed to the improved performance by making the model more robust to variations in the input data.

Overall, achieving 87% accuracy on CIFAR-10 is a strong result. The revised model architecture and data augmentation strategies have significantly improved the model's performance. Continue experimenting with the mentioned techniques to push the model's accuracy even higher.